### PR TITLE
dxTagBox: Fix low performance when a large amount of data is selected or deselected (T584740)

### PR DESCRIPTION
**Pull request by vconst**

### CODE REVIEW

File: js/data/array_store.js [hunk 1]
```diff
@@ -76,8 +76,8 @@ var ArrayStore = Store.inherit({
         });
     },
 
-    _byKeyImpl: function(key) {
-        var index = this._indexByKey(key);
+    _byKeyImpl: function(key, extraOptions, cache) {
```
Comment (jsdmitry) [inline js/data/array_store.js]: Where the extraOptions parameter use?

**Comment by AlekseyMartynov** [comment]
😲💩

P.S. 
- IMO, "cache" is the wrong name. If I understand correctly, this PR implements what behaves as [index](https://en.wikipedia.org/wiki/Database_index).
- Why pass it in parameters? Should be possible to store in `this`.
- Index needs to be invalidated on update/insert/delete.
- Tag box and large data... 🤔

P.P.S. "should works fast" → "should use index"

File: js/data/array_store.js (hunk 1 above)
Comment (AlekseyMartynov) [reply js/data/array_store.js]: https://github.com/DevExpress/DevExtreme/search?q=extraOptions&type=Code

Comment (jsdmitry) [reply js/data/array_store.js]: The extraOptions variable is unused in this method.

Comment (vconst) [reply js/data/array_store.js]: But extraOptions parameter is used in ODataStore.

**Comment by vconst** [comment]
@AlekseyMartynov 
Currently, the dxTagBox call byKey for each selected item and It is impossible to change this behavior without BC. So I add cache to improve performance in 17.2 without any breaking changes. But in 18.1 we can try to remove byKey and use load.

>> IMO, "cache" is the wrong name. If I understand correctly, this PR implements what behaves as index.

I use the cache name for this variable because it is a temporary data that improves access to items by keys. And it is not important that indexByKey object is created in this cache.

>> Why pass it in parameters? Should be possible to store in this.

>> Index needs to be invalidated on update/insert/delete.

Cache variable is the object that shared between several byKey calls. This variable is disposed when byKey calls are ended and there are no problems with cache invalidation on update/insert/delete. If the cache will be stored in `this` then cache invalidation problem will be very very actual.

>>Tag box and large data...
Dashboard and Report teams think in another way.

>> "should works fast" → "should use index"
I will rename this tests

**Comment by AlekseyMartynov** [comment]
I think we can come up with a better modular API for that.

```js
if("setKeyIndex" in store)
  store.setKeyIndex(new KeyIndex())

// selection

if("setKeyIndex" in store)
  store.setKeyIndex(null);
```

I can work on it when I return.

**Comment by vconst** [comment]
💩
Why is adding setKeyIndex method better when adding argument? I think that functional approach with one additional argument is simpler, shorter and cleaner.

**Comment by AlekseyMartynov** [comment]
Two reasons:
- This new code is a workaround for a specific widget-store pair. It seems better if it's independent and pluggable.
- Function parameters become awkward as noted by @jsdmitry

**Comment by vconst** [comment]
Cache from ArrayStore was removed

**Review by jsdmitry — DISMISSED** [review]

**Comment by AlekseyMartynov** [comment]
@jsdmitry http://www.bokudjava.ru/S_30.html

**Review by jsdmitry — APPROVED** [review]

**Pull request merged**